### PR TITLE
Replace total active job count limit with a single batch job limit

### PR DESCRIPTION
**Pull request by yuzhu**
### What changes are proposed in this pull request?

Before this change, the replication checker issues O(n) rpc calls to check the status of every job it submitted, resulting in a large delay.
Instead of limiting the number of active job outstanding submitted by the replication checker, we will limit the number of jobs submitted in one iteration.  This is a temporary solution until proper accounting and resource isolation is implemented on the job service. 

### Why are the changes needed?
Free replication checker from doing many rpcs and not being able to make progress

### Does this PR introduce any user facing changes?
No

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java [hunk 1]
```diff
@@ -137,20 +130,6 @@ public void heartbeat() throws InterruptedException {
       return;
     }
 
-    final Iterator<Long> jobIterator = mActiveJobToInodeID.keySet().iterator();
-
-    while (jobIterator.hasNext()) {
```
Comment (yuzhu) [inline core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:142]: This loop used to take forever and blocks replication checker essentially.

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java [hunk 2]
```diff
@@ -219,17 +199,15 @@ public void close() {
 
   private void checkMisreplicated(Set<Long> inodes, ReplicationHandler handler)
       throws InterruptedException {
+    int singleBatch = 0;
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:202]: I don't see this local variable got updated.

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java [hunk 3]
```diff
@@ -273,18 +249,16 @@ private void checkMisreplicated(Set<Long> inodes, ReplicationHandler handler)
   private Set<Long> check(Set<Long> inodes, ReplicationHandler handler, Mode mode)
       throws InterruptedException {
     Set<Long> processedFileIds = new HashSet<>();
+    int singleBatch = 0;
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:252]: same here, how is this local variable got updated?

**Review by apc999 — CHANGES_REQUESTED** [review]
is this PR ready to review?

**Comment by yuzhu** [comment]
@apc999 oh wow, clearly it was not ready for review. made the change in a hurry, now fixed.

**Comment by ggezer** [comment]
@yuzhu @apc999 Is batch query capability in the plans for job service?

**Comment by yuzhu** [comment]
> @yuzhu @apc999 Is batch query capability in the plans for job service?
 
In general, we want to improve the visibility of job status and allow filtering of jobs so that the user/client can get back more useful job status information rather than a dump of every job in the system.  The initial idea is to introduce group id for jobs so that we can find all jobs in a particular group and associate that with a command that the user issued  (like a distributedLoad) or a background task(like replication checker), but welcome feedback.

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java [hunk 4]
```diff
@@ -252,9 +228,11 @@ private void checkMisreplicated(Set<Long> inodes, ReplicationHandler handler)
...
               : findMisplacedBlock(file, blockInfo).entrySet()) {
             try {
-              final long jobId =
-                  handler.migrate(inodePath.getUri(), blockId, entry.getKey(), entry.getValue());
-              mActiveJobToInodeID.put(jobId, inodeId);
+              if (mRequestBatchCount >= mMaxSingleBatch) {
+                return;
+              }
+              handler.migrate(inodePath.getUri(), blockId, entry.getKey(), entry.getValue());
+              mRequestBatchCount++;
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:235]: is it possible to have race here?

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java [hunk 5]
```diff
@@ -220,12 +202,6 @@ public void close() {
   private void checkMisreplicated(Set<Long> inodes, ReplicationHandler handler)
       throws InterruptedException {
     for (long inodeId : inodes) {
-      if (mActiveJobToInodeID.size() >= mMaxActiveJobs) {
-        return;
-      }
-      if (mActiveJobToInodeID.containsValue(inodeId)) {
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:226]: don't we check duplicated job any more?

**Review by apc999 — CHANGES_REQUESTED** [review]

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java (hunk 4 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:235]: Replication checker is currently single threaded. Will need to consider race when we multi thread it.

File: core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java (hunk 5 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java:226]: Unfortunately we.do need to. And given the current job master interface, that means checking each job for completion status to maintain the active job list. I will try to see if we can make a smaller change.here without changing job master interface, but that may not be possible

**Comment by yuzhu** [comment]
addressing this issue in #13831 instead

**Pull request closed without merging**